src/python/pants/backend/scala/compile/scalac_plugins.py

```diff
@@ -107,6 +165,94 @@ async def global_scalac_plugins(
 
 
+@rule
+async def all_scala_plugin_targets(targets: AllTargets) -> AllScalaPluginTargets:
+    return AllScalaPluginTargets(
+        tgt for tgt in targets if tgt.has_fields((ScalacPluginArtifactField, ScalacPluginNameField))
+    )
+
+
+@rule
+async def resolve_scala_plugins_for_target(
+    request: ScalaPluginsForTargetRequest,
+    all_scala_plugins: AllScalaPluginTargets,
+    jvm: JvmSubsystem,
+    scalac: Scalac,
+) -> ScalaPluginTargetsForTarget:
+
+    target = request.target
+    resolve = target[JvmResolveField].normalized_value(jvm)
+
+    plugin_names = target.get(ScalaConsumedPluginNamesField).value
+    if plugin_names is None:
+        plugin_names_by_resolve = scalac.parsed_default_plugins()
+        plugin_names = tuple(plugin_names_by_resolve.get(resolve) or ())
+
+    candidate_plugins: list[Target] = []
+    for plugin in all_scala_plugins:
+        if _plugin_name(plugin) in plugin_names:
+            candidate_plugins.append(plugin)
+
+    artifact_address_inputs = (
+        plugin[ScalacPluginArtifactField].value for plugin in candidate_plugins
+    )
+
+    artifact_addresses = await MultiGet(
+        # `is not None` is solely to satiate mypy. artifact field is required.
+        Get(Address, AddressInput, AddressInput.parse(ai))
+        for ai in artifact_address_inputs
+        if ai is not None
+    )
+
+    candidate_artifacts = await Get(Targets, Addresses(artifact_addresses))
+
+    plugins: dict[str, tuple[Target, Target]] = {}  # Maps plugin name to relevant JVM artifact
+    for plugin, artifact in zip(candidate_plugins, candidate_artifacts):
+        if artifact[JvmResolveField].normalized_value(jvm) != resolve:
+            continue
+
+        plugins[_plugin_name(plugin)] = (plugin, artifact)
+
+    for plugin_name in plugin_names:
+        if plugin_name not in plugins:
+            raise Exception(f"Could not find Scala plugin `{plugin_name}` in resolve `{resolve}`")
+
+    plugin_targets, artifact_targets = zip(*plugins.values()) if plugins else ((), ())
+    return ScalaPluginTargetsForTarget(Targets(plugin_targets), Targets(artifact_targets))
+
+
+def _plugin_name(target: Target) -> str:
+    return target[ScalacPluginNameField].value or target.address.target_name
+
+
+@rule
+async def fetch_plugins(request: ScalaPluginsRequest) -> ScalaPlugins:
+
+    # Fetch all the artifacts
+    coarsened_targets = await Get(
+        CoarsenedTargets, Addresses(target.address for target in request.artifacts)
+    )
+    fallible_artifacts = await MultiGet(
+        Get(
+            FallibleClasspathEntry,
+            CoursierFetchRequest(ct, resolve=request.resolve),
+        )
+        for ct in coarsened_targets
+    )
+
+    artifacts = FallibleClasspathEntry.if_all_succeeded(fallible_artifacts)
+    if artifacts is None:
+        failed = [i for i in fallible_artifacts if i.exit_code != 0]
+        raise Exception(f"Something went wrong! {failed=}")
```

Comment:
This line triggered a syntax error in CI due to the `failed=` syntax. Also maybe add a more informative error message?